### DISCUSSION TITLE
How to apply query filters globally?

### DISCUSSION THREAD

**Discussion by TheShubhendra** [sqlalchemy/sqlalchemy#11779] [Usage Questions] [answered]
I have many models that inherits following `VisibilityMixin` class with the `sqlalchemy.orm.DeclarativeBase` class
```python
class VisibilityMixin:
    is_visible = Column(Boolean, server_default="1")
    is_deleted = Column(Boolean, server_default="0")


```
Currently I am filtering those model objects in every endpoints based on these 2 attributes 
```python
@router.get(
    "/my",
    response_model=list[schemas.Category],
    description="Get all subscribed courses",
)
async def get_my_categories(
    db: Session = Depends(get_db),
    course_ids: list[int] = Depends(get_course_ids),
    role: Role = Depends(get_role),
):
    categories = (
        db.query(Category)
        .join(Category.courses)
        .filter(
            and_(
                Course.id.in_(course_ids),
                Course.is_deleted.is_(False),
                Course.is_visible.is_(True) if role != Role.ADMIN else True,
                Category.is_deleted.is_(False),
                Category.is_visible.is_(True) if role != Role.ADMIN else True,
            )
        )
        .order_by(Category.sequence)
        .all()
    )

    return categories
```
What I am looking is any simple way to globally apply these filters based of role or some attribute. May be I can create a subclass `sqlalchemy.orm.Query` and apply the filters there. But how can I do it using new session.execute API as it also support asynchronous queries ?

**Comment by zzzeek** [comment, marked as answer]
SQLAlchemy has full support for this feature, though the documentation for it does not have a prominent top-level section it seems so I dont blame you for not seeing it, you would start with the [with_loader_criteria](https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#sqlalchemy.orm.with_loader_criteria) option that will add the criteria you want, then typically this is combined with [do_orm_execute](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.do_orm_execute) to automate it.   two working examples can then be seen at https://docs.sqlalchemy.org/en/20/orm/examples.html#module-examples.extending_query

**Reply by TheShubhendra** [reply]
Thanks @zzzeek This should work! Just a small confusion https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#sqlalchemy.orm.with_loader_criteria:~:text=%40event.listens_for(%22do_orm_execute%22%2C%20session) in this example I think first parameter should be `session` then `do_orm_execute`. I thought about creating a issue but just wanted to confirm here first.

**Reply by zzzeek** [reply]
there's two functions.  one is the with_loader_criteria() ORM option that you linked.  then the other is the do_orm_execute() event hook, which receives just one argument, which is an object called [ORMExecuteState](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.ORMExecuteState) .   this object has lots of methods and accessors on it to tell you everything we can about the query that's being run, including how to get at the current session which is the [session](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.ORMExecuteState.session) accessor.

I would recommend running the example script to play with it:  https://github.com/sqlalchemy/sqlalchemy/blob/main/examples/extending_query/filter_public.py

**Comment by TheShubhendra** [comment]
For anyone who come across the similar issue, I have written an article on application of global filters that You might find helpful: [Role-Based Row Filtering: Advanced SQLAlchemy Techniques](https://theshubhendra.medium.com/role-based-row-filtering-advanced-sqlalchemy-techniques-733e6b1328f6). I thought I’d share this since I couldn’t find a solution or any ideas anywhere on web when I was searching for answers previously.
